### ISSUE TITLE
Memory exhaustion upon inspection of a newly created object

### ISSUE THREAD

**Issue by JohnKiro** [pharo-project/pharo-vm#848]
Hello Pharo Community,

I've been exploring Pharo recently for over 2 weeks now, and so far I'm impressed (I'm also new to the Smalltalk world). However, earlier today, I faced a nasty bug for the first time. The good news is that I could reproduce it, and I hope the description I give here will be sufficient for the Pharo team to investigate.

What I did:

1. Created a seemingly innocent class (like I did, several times before). The class code is attached.
2. In the Playground, try inspect an instance of the class, by typing `CharRange new`, and then pressing **Ctrl+I**.

At this point, the Inspector window is openend, but **Pharo memory usage starts to grow without limit**, until I kill the process!

I tested using the **stable Pharo 12 image**, and I could reproduce it with a fresh image, by filing-in the code. In the screenshot below, the Pharo image is taking 2GB of RAM, and it keeps growing, until I kill it.

The CharRange class code is attached. Please let me know if you have any comments or questions.

### More Information:

- Reproduced also on **Pharo 13** (latest VM snapshot).
- Not reproduced with _doing_ (Ctrl+D) or _printing_ (Ctrl+P). So **bug probably located in the Inspector**. I wish I was more experienced to try to debug it, but I'm not there yet :)

Thanks.
John

![image](https://github.com/user-attachments/assets/632b8e6e-d46f-4360-8efd-701fb7387234)
[CharRange.st.zip](https://github.com/user-attachments/files/16790214/CharRange.st.zip)

**Comment by guillep** [comment]
Hi @JohnKiro , thanks for coming here! I hope you're having fun! :)

I've checked your code, and it seems you have been bitten by an infinite recursion.
If you look at your code

```smalltalk
= anObject
   "Two ranges are equal if they have same start and end."

   ^ self = anObject or: (start = anObject start and: end = anObject end)! !
```

Your = method will call itself over and over when doing `self = anObject`

Now, explaining why this takes a lot of memory would take me a lot of time.
I did lightning talk last ESUG this year about this topic. 
I attach the PDF in case you're interested in understanding WHY this happens.
Although, my slides are not annotated, and without my fabulous (?) comedian skills to tell the story they would be difficult to follow.

[ESUG-STACK-SHOW-PROJECT.pdf](https://github.com/user-attachments/files/16795610/ESUG-STACK-SHOW-PROJECT.pdf)

TL;DR; Pharo supports virtually infinite stacks, which create problems with very loooooong recursions. We are integrating stack overflow support in the upcoming releases, see https://github.com/pharo-project/pharo-vm/pull/710, integrated yesterday...

**Comment by JohnKiro** [comment]
Hello @guillep 

Thank you so much! It's indeed a lot of fun to explore Pharo. Smalltalk/Pharo syntax on a postcard can easily attract anyone interested in programming :) The live environment is also great to work with!

About this issue, now it's clear... I should have used `==` instead, and I didn't suspect a stack overflow. It's great that you're implementing a prevention for this already :)

And thanks for the link to the slides! I'll check out the presentation itself on youtube or Pharo site. It'll be interesting to watch for sure.

Thanks again.

P.S. pls let me know if I need to **close** this ticket.

**Comment by JohnKiro** [comment]
As Indicated in Guille's comment, stack overflow handling is being introduced (see https://github.com/pharo-project/pharo-vm/pull/710), so as to avoid uncontrolled VM size growth.
